### PR TITLE
A new public API for getting locations of PSModulePath elements

### PR DESCRIPTION
**Pull request by JamesWTruher**
# PR Summary

A new _public_ API for retrieving the PSModulePath element for PSHOME, User, and Shared locations.

## PR Context

Retrieving the locations for the possible locations for portions of the Module path have been inconsistent. This is a new public api which enables 3rd party tools to have a consistent way to retrieve this information. It should also allow PowerShellGet to simplify code for installations in PowerShell 7.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 1]
```diff
@@ -1172,6 +1172,42 @@ private static string AddToPath(string basePath, string pathToAdd, int insertPos
             return result.ToString();
         }
 
+        /// <summary>
+        /// The available module path types.
+        /// </summary>
+        public enum ModulePathType
```
Comment (ThomasNieto) [inline src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:1178]: Should this be PSModulePathType to match the method?

Comment (JamesWTruher) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:1178]: sure

File: test/powershell/engine/Module/ModulePath.Tests.ps1 [hunk 2]
```diff
@@ -188,3 +188,53 @@ Describe "SxS Module Path Basic Tests" -tags "CI" {
...
+        # resolve the paths to ensure they are in the correct format
+        $currentModulePathElements = $env:PSModulePath -split [System.IO.Path]::PathSeparator | Foreach-Object { (Resolve-Path $_).Path }
+    }
+
+    AfterAll {
+        Unregister-PSRepository -Name $localSourceName -ErrorAction SilentlyContinue
+    }
+
+    It "The GetPSModulePath api is available" {
+        { [System.Management.Automation.ModuleIntrinsics]::GetPSModulePath("PSHOME") } | Should -not -throw
```
Comment (SteveL-MSFT) [inline test/powershell/engine/Module/ModulePath.Tests.ps1:222]: I don't think this specific test case is needed since it's implicitly covered below.

**Review by SteveL-MSFT — APPROVED** [review]
One suggestion

File: test/powershell/engine/Module/assets/PowerShell.TestPackage.3.2.1.nupkg
Comment (TravisEz13) [inline test/powershell/engine/Module/assets/PowerShell.TestPackage.3.2.1.nupkg:1]: Is this here by mistake?

File: test/powershell/engine/Module/ModulePath.Tests.ps1 (hunk 2 above)
Comment (JamesWTruher) [reply test/powershell/engine/Module/ModulePath.Tests.ps1:222]: removed

File: test/powershell/engine/Module/assets/PowerShell.TestPackage.3.2.1.nupkg
Comment (JamesWTruher) [reply test/powershell/engine/Module/assets/PowerShell.TestPackage.3.2.1.nupkg:1]: no - it's used by the tests (line 199)

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 3]
```diff
@@ -1172,6 +1172,42 @@ private static string AddToPath(string basePath, string pathToAdd, int insertPos
...
+            /// <summary>The shared module path.</summary>
+            Shared
+        }
+
+        /// <summary>
+        /// Retrieve the current PSModulePath of the specified type.
+        /// </summary>
+        /// <param name="type">The type of module path to retrieve. This can be User, PSHome, or Shared.</param>
+        /// <returns>The string representing the requested module path type.</returns>
+        public static string GetPSModulePath(PSModulePathType type)
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:1195]: I am wondering if the parameter name should be `scope` to be consistent with other cmdlets and PowerShellGet.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 4]
```diff
@@ -1172,6 +1172,42 @@ private static string AddToPath(string basePath, string pathToAdd, int insertPos
...
+        public enum PSModulePathType
+        {
+            /// <summary>The users module path.</summary>
+            User,
+
+            /// <summary>The PSHOME module path.</summary>
+            PSHome,
+
+            /// <summary>The shared module path.</summary>
+            Shared
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs]: Resolve as feedback was addressed.

File: src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs [hunk 5]
```diff
@@ -1172,6 +1172,42 @@ private static string AddToPath(string basePath, string pathToAdd, int insertPos
...
+        /// <summary>
+        /// The available module path types.
+        /// </summary>
+        public enum PSModulePathType
+        {
+            /// <summary>The users module path.</summary>
+            User,
+
+            /// <summary>The PSHOME module path.</summary>
+            PSHome,
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs]: Resolve as feedback was addressed.

**Pull request merged**